// @ariakit/test/cjs/react.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true}); function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { newObj[key] = obj[key]; } } } newObj.default = obj; return newObj; } }"use client";
require('./__chunks/I2BROIQW.cjs');


var _SEWCIANIcjs = require('./__chunks/SEWCIANI.cjs');


var _SDUGTL2Ncjs = require('./__chunks/SDUGTL2N.cjs');


var _PJWPDIUXcjs = require('./__chunks/PJWPDIUX.cjs');


var _DAHHEUOUcjs = require('./__chunks/DAHHEUOU.cjs');


var _747HORQZcjs = require('./__chunks/747HORQZ.cjs');


var _T5KVOIJTcjs = require('./__chunks/T5KVOIJT.cjs');


var _LYOKT276cjs = require('./__chunks/LYOKT276.cjs');



var _2KXSRUIVcjs = require('./__chunks/2KXSRUIV.cjs');


var _ESE7QHAQcjs = require('./__chunks/ESE7QHAQ.cjs');


var _VZINRVVHcjs = require('./__chunks/VZINRVVH.cjs');


var _KVNENCXTcjs = require('./__chunks/KVNENCXT.cjs');


var _A5IFXY4Jcjs = require('./__chunks/A5IFXY4J.cjs');


var _FU2U6RZLcjs = require('./__chunks/FU2U6RZL.cjs');


var _NIEATM7Rcjs = require('./__chunks/NIEATM7R.cjs');



var _ZMMVJ3OTcjs = require('./__chunks/ZMMVJ3OT.cjs');



var _OVN5OYWKcjs = require('./__chunks/OVN5OYWK.cjs');

// src/react.tsx
var _react = require('react');
var _react3 = require('@testing-library/react'); var ReactTestingLibrary = _interopRequireWildcard(_react3);
var _jsxruntime = require('react/jsx-runtime');
async function render2(ui, options) {
  const wrapper = (props) => {
    const Wrapper = options == null ? void 0 : options.wrapper;
    const element = Wrapper ? /* @__PURE__ */ _jsxruntime.jsx.call(void 0, Wrapper, _OVN5OYWKcjs.__spreadValues.call(void 0, {}, props)) : props.children;
    if (!(options == null ? void 0 : options.strictMode))
      return element;
    return /* @__PURE__ */ _jsxruntime.jsx.call(void 0, _react.StrictMode, { children: element });
  };
  return _ZMMVJ3OTcjs.wrapAsync.call(void 0, async () => {
    const { unmount } = ReactTestingLibrary.render(ui, _OVN5OYWKcjs.__spreadProps.call(void 0, _OVN5OYWKcjs.__spreadValues.call(void 0, {}, options), { wrapper }));
    await _ZMMVJ3OTcjs.nextFrame.call(void 0, );
    return unmount;
  });
}

















exports.blur = _KVNENCXTcjs.blur; exports.click = _PJWPDIUXcjs.click; exports.dispatch = _FU2U6RZLcjs.dispatch; exports.focus = _A5IFXY4Jcjs.focus; exports.hover = _LYOKT276cjs.hover; exports.mouseDown = _T5KVOIJTcjs.mouseDown; exports.mouseUp = _747HORQZcjs.mouseUp; exports.press = _ESE7QHAQcjs.press; exports.q = _2KXSRUIVcjs.q; exports.query = _2KXSRUIVcjs.query; exports.render = render2; exports.select = _DAHHEUOUcjs.select; exports.sleep = _NIEATM7Rcjs.sleep; exports.tap = _SDUGTL2Ncjs.tap; exports.type = _VZINRVVHcjs.type; exports.waitFor = _SEWCIANIcjs.waitFor;
