// @ariakit/core/cjs/utils/misc.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});"use client";





















var _LMT335NJcjs = require('../__chunks/LMT335NJ.cjs');
require('../__chunks/WYXAWBLE.cjs');





















exports.afterPaint = _LMT335NJcjs.afterPaint; exports.applyState = _LMT335NJcjs.applyState; exports.beforePaint = _LMT335NJcjs.beforePaint; exports.chain = _LMT335NJcjs.chain; exports.cx = _LMT335NJcjs.cx; exports.defaultValue = _LMT335NJcjs.defaultValue; exports.disabledFromProps = _LMT335NJcjs.disabledFromProps; exports.getKeys = _LMT335NJcjs.getKeys; exports.hasOwnProperty = _LMT335NJcjs.hasOwnProperty; exports.identity = _LMT335NJcjs.identity; exports.invariant = _LMT335NJcjs.invariant; exports.isEmpty = _LMT335NJcjs.isEmpty; exports.isFalsyBooleanCallback = _LMT335NJcjs.isFalsyBooleanCallback; exports.isInteger = _LMT335NJcjs.isInteger; exports.isObject = _LMT335NJcjs.isObject; exports.noop = _LMT335NJcjs.noop; exports.normalizeString = _LMT335NJcjs.normalizeString; exports.omit = _LMT335NJcjs.omit; exports.pick = _LMT335NJcjs.pick; exports.shallowEqual = _LMT335NJcjs.shallowEqual;
