// @ariakit/react-core/cjs/__chunks/2UXXG7NZ.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});"use client";


var _FD3VCGWPcjs = require('./FD3VCGWP.cjs');


var _H2TTW65Qcjs = require('./H2TTW65Q.cjs');




var _VKYTJKYNcjs = require('./VKYTJKYN.cjs');


var _WLJVTO7Acjs = require('./WLJVTO7A.cjs');






var _JTVNRMACcjs = require('./JTVNRMAC.cjs');




var _WYXAWBLEcjs = require('./WYXAWBLE.cjs');

// src/radio/radio.ts
var _react = require('react');
var _misc = require('@ariakit/core/utils/misc');
function getIsChecked(value, storeValue) {
  if (storeValue === void 0)
    return;
  if (value != null && storeValue != null) {
    return storeValue === value;
  }
  return !!storeValue;
}
function isNativeRadio(tagName, type) {
  return tagName === "input" && (!type || type === "radio");
}
var useRadio = _VKYTJKYNcjs.createHook.call(void 0, 
  (_a) => {
    var _b = _a, { store, name, value, checked } = _b, props = _WYXAWBLEcjs.__objRest.call(void 0, _b, ["store", "name", "value", "checked"]);
    const context = _FD3VCGWPcjs.useRadioContext.call(void 0, );
    store = store || context;
    const id = _JTVNRMACcjs.useId.call(void 0, props.id);
    const ref = _react.useRef.call(void 0, null);
    const isChecked = _WLJVTO7Acjs.useStoreState.call(void 0, 
      store,
      (state) => checked != null ? checked : getIsChecked(value, state == null ? void 0 : state.value)
    );
    _react.useEffect.call(void 0, () => {
      if (!id)
        return;
      if (!isChecked)
        return;
      const isActiveItem = (store == null ? void 0 : store.getState().activeId) === id;
      if (isActiveItem)
        return;
      store == null ? void 0 : store.setActiveId(id);
    }, [store, isChecked, id]);
    const onChangeProp = props.onChange;
    const tagName = _JTVNRMACcjs.useTagName.call(void 0, ref, props.as || "input");
    const nativeRadio = isNativeRadio(tagName, props.type);
    const disabled = _misc.disabledFromProps.call(void 0, props);
    const [propertyUpdated, schedulePropertyUpdate] = _JTVNRMACcjs.useForceUpdate.call(void 0, );
    _react.useEffect.call(void 0, () => {
      const element = ref.current;
      if (!element)
        return;
      if (nativeRadio)
        return;
      if (isChecked !== void 0) {
        element.checked = isChecked;
      }
      if (name !== void 0) {
        element.name = name;
      }
      if (value !== void 0) {
        element.value = `${value}`;
      }
    }, [propertyUpdated, nativeRadio, isChecked, name, value]);
    const onChange = _JTVNRMACcjs.useEvent.call(void 0, (event) => {
      if (disabled) {
        event.preventDefault();
        event.stopPropagation();
        return;
      }
      if (!nativeRadio) {
        event.currentTarget.checked = true;
        schedulePropertyUpdate();
      }
      onChangeProp == null ? void 0 : onChangeProp(event);
      if (event.defaultPrevented)
        return;
      store == null ? void 0 : store.setValue(value);
    });
    const onClickProp = props.onClick;
    const onClick = _JTVNRMACcjs.useEvent.call(void 0, (event) => {
      onClickProp == null ? void 0 : onClickProp(event);
      if (event.defaultPrevented)
        return;
      if (nativeRadio)
        return;
      onChange(event);
    });
    const onFocusProp = props.onFocus;
    const onFocus = _JTVNRMACcjs.useEvent.call(void 0, (event) => {
      onFocusProp == null ? void 0 : onFocusProp(event);
      if (event.defaultPrevented)
        return;
      if (!nativeRadio)
        return;
      if (!store)
        return;
      const { moves, activeId } = store.getState();
      if (!moves)
        return;
      if (id && activeId !== id)
        return;
      onChange(event);
    });
    props = _WYXAWBLEcjs.__spreadProps.call(void 0, _WYXAWBLEcjs.__spreadValues.call(void 0, {
      id,
      role: !nativeRadio ? "radio" : void 0,
      type: nativeRadio ? "radio" : void 0,
      "aria-checked": isChecked
    }, props), {
      ref: _JTVNRMACcjs.useMergeRefs.call(void 0, ref, props.ref),
      onChange,
      onClick,
      onFocus
    });
    props = _H2TTW65Qcjs.useCompositeItem.call(void 0, _WYXAWBLEcjs.__spreadValues.call(void 0, { store, clickOnEnter: !nativeRadio }, props));
    return _WYXAWBLEcjs.__spreadValues.call(void 0, {
      name: nativeRadio ? name : void 0,
      value: nativeRadio ? value : void 0,
      checked: isChecked
    }, props);
  }
);
var Radio = _VKYTJKYNcjs.createMemoComponent.call(void 0, (props) => {
  const htmlProps = useRadio(props);
  return _VKYTJKYNcjs.createElement.call(void 0, "input", htmlProps);
});
if (process.env.NODE_ENV !== "production") {
  Radio.displayName = "Radio";
}




exports.useRadio = useRadio; exports.Radio = Radio;
