// @ariakit/react-core/cjs/__chunks/FQGSRSYD.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});"use client";



var _JVORNAMDcjs = require('./JVORNAMD.cjs');


var _VKYTJKYNcjs = require('./VKYTJKYN.cjs');

// src/composite/composite-context.tsx
var _react = require('react');
var ctx = _VKYTJKYNcjs.createStoreContext.call(void 0, 
  [_JVORNAMDcjs.CollectionContextProvider],
  [_JVORNAMDcjs.CollectionScopedContextProvider]
);
var useCompositeContext = ctx.useContext;
var useCompositeScopedContext = ctx.useScopedContext;
var useCompositeProviderContext = ctx.useProviderContext;
var CompositeContextProvider = ctx.ContextProvider;
var CompositeScopedContextProvider = ctx.ScopedContextProvider;
var CompositeItemContext = _react.createContext.call(void 0, 
  void 0
);
var CompositeRowContext = _react.createContext.call(void 0, 
  void 0
);









exports.useCompositeContext = useCompositeContext; exports.useCompositeScopedContext = useCompositeScopedContext; exports.useCompositeProviderContext = useCompositeProviderContext; exports.CompositeContextProvider = CompositeContextProvider; exports.CompositeScopedContextProvider = CompositeScopedContextProvider; exports.CompositeItemContext = CompositeItemContext; exports.CompositeRowContext = CompositeRowContext;
