// @ariakit/react-core/cjs/__chunks/KJQSD4SL.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});"use client";


var _P34T42GIcjs = require('./P34T42GI.cjs');


var _HY2NPBHWcjs = require('./HY2NPBHW.cjs');




var _VKYTJKYNcjs = require('./VKYTJKYN.cjs');





var _JTVNRMACcjs = require('./JTVNRMAC.cjs');




var _WYXAWBLEcjs = require('./WYXAWBLE.cjs');

// src/form/form-field.ts
var _react = require('react');
var _dom = require('@ariakit/core/utils/dom');
var _misc = require('@ariakit/core/utils/misc');
function getNamedElement(ref, name) {
  const element = ref.current;
  if (!element)
    return null;
  if (element.name === name)
    return element;
  if (element.form) {
    return element.form.elements.namedItem(name);
  }
  const document = _dom.getDocument.call(void 0, element);
  return document.getElementsByName(name)[0];
}
function useItem(store, name, type) {
  return store.useState(
    (state) => state.items.find((item) => item.type === type && item.name === name)
  );
}
var useFormField = _VKYTJKYNcjs.createHook.call(void 0, 
  (_a) => {
    var _b = _a, {
      store,
      name: nameProp,
      getItem: getItemProp,
      touchOnBlur = true
    } = _b, props = _WYXAWBLEcjs.__objRest.call(void 0, _b, [
      "store",
      "name",
      "getItem",
      "touchOnBlur"
    ]);
    const context = _P34T42GIcjs.useFormContext.call(void 0, );
    store = store || context;
    _misc.invariant.call(void 0, 
      store,
      process.env.NODE_ENV !== "production" && "FormField must be wrapped in a Form component."
    );
    const name = `${nameProp}`;
    const id = _JTVNRMACcjs.useId.call(void 0, props.id);
    const ref = _react.useRef.call(void 0, null);
    store.useValidate(async () => {
      const element = getNamedElement(ref, name);
      if (!element)
        return;
      await Promise.resolve();
      if ("validity" in element && !element.validity.valid) {
        store == null ? void 0 : store.setError(name, element.validationMessage);
      }
    });
    const getItem = _react.useCallback.call(void 0, 
      (item) => {
        const nextItem = _WYXAWBLEcjs.__spreadProps.call(void 0, _WYXAWBLEcjs.__spreadValues.call(void 0, {}, item), { id: id || item.id, name, type: "field" });
        if (getItemProp) {
          return getItemProp(nextItem);
        }
        return nextItem;
      },
      [id, name, getItemProp]
    );
    const onBlurProp = props.onBlur;
    const touchOnBlurProp = _JTVNRMACcjs.useBooleanEvent.call(void 0, touchOnBlur);
    const onBlur = _JTVNRMACcjs.useEvent.call(void 0, (event) => {
      onBlurProp == null ? void 0 : onBlurProp(event);
      if (event.defaultPrevented)
        return;
      if (!touchOnBlurProp(event))
        return;
      store == null ? void 0 : store.setFieldTouched(name, true);
    });
    const label = useItem(store, name, "label");
    const error = useItem(store, name, "error");
    const description = useItem(store, name, "description");
    const describedBy = _misc.cx.call(void 0, 
      error == null ? void 0 : error.id,
      description == null ? void 0 : description.id,
      props["aria-describedby"]
    );
    const invalid = store.useState(
      () => !!(store == null ? void 0 : store.getError(name)) && store.getFieldTouched(name)
    );
    props = _WYXAWBLEcjs.__spreadProps.call(void 0, _WYXAWBLEcjs.__spreadValues.call(void 0, {
      id,
      "aria-labelledby": label == null ? void 0 : label.id,
      "aria-invalid": invalid
    }, props), {
      "aria-describedby": describedBy || void 0,
      ref: _JTVNRMACcjs.useMergeRefs.call(void 0, ref, props.ref),
      onBlur
    });
    props = _HY2NPBHWcjs.useCollectionItem.call(void 0, _WYXAWBLEcjs.__spreadProps.call(void 0, _WYXAWBLEcjs.__spreadValues.call(void 0, { store }, props), { name, getItem }));
    return props;
  }
);
var FormField = _VKYTJKYNcjs.createMemoComponent.call(void 0, (props) => {
  const htmlProps = useFormField(props);
  return _VKYTJKYNcjs.createElement.call(void 0, "input", htmlProps);
});
if (process.env.NODE_ENV !== "production") {
  FormField.displayName = "FormField";
}




exports.useFormField = useFormField; exports.FormField = FormField;
