// @ariakit/react-core/cjs/__chunks/KKBPJ4X4.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});"use client";


var _XCM6Y3AXcjs = require('./XCM6Y3AX.cjs');


var _QKKT6RXXcjs = require('./QKKT6RXX.cjs');




var _VKYTJKYNcjs = require('./VKYTJKYN.cjs');


var _JTVNRMACcjs = require('./JTVNRMAC.cjs');




var _WYXAWBLEcjs = require('./WYXAWBLE.cjs');

// src/dialog/dialog-dismiss.tsx
var _react = require('react');
var _jsxruntime = require('react/jsx-runtime');
var useDialogDismiss = _VKYTJKYNcjs.createHook.call(void 0, 
  (_a) => {
    var _b = _a, { store } = _b, props = _WYXAWBLEcjs.__objRest.call(void 0, _b, ["store"]);
    const context = _QKKT6RXXcjs.useDialogScopedContext.call(void 0, );
    store = store || context;
    const onClickProp = props.onClick;
    const onClick = _JTVNRMACcjs.useEvent.call(void 0, (event) => {
      onClickProp == null ? void 0 : onClickProp(event);
      if (event.defaultPrevented)
        return;
      store == null ? void 0 : store.hide();
    });
    const children = _react.useMemo.call(void 0, 
      () => /* @__PURE__ */ _jsxruntime.jsxs.call(void 0, 
        "svg",
        {
          "aria-label": "Dismiss popup",
          display: "block",
          fill: "none",
          stroke: "currentColor",
          strokeLinecap: "round",
          strokeLinejoin: "round",
          strokeWidth: "1.5pt",
          viewBox: "0 0 16 16",
          height: "1em",
          width: "1em",
          children: [
            /* @__PURE__ */ _jsxruntime.jsx.call(void 0, "line", { x1: "4", y1: "4", x2: "12", y2: "12" }),
            /* @__PURE__ */ _jsxruntime.jsx.call(void 0, "line", { x1: "4", y1: "12", x2: "12", y2: "4" })
          ]
        }
      ),
      []
    );
    props = _WYXAWBLEcjs.__spreadProps.call(void 0, _WYXAWBLEcjs.__spreadValues.call(void 0, {
      "data-dialog-dismiss": "",
      children
    }, props), {
      onClick
    });
    props = _XCM6Y3AXcjs.useButton.call(void 0, props);
    return props;
  }
);
var DialogDismiss = _VKYTJKYNcjs.createComponent.call(void 0, (props) => {
  const htmlProps = useDialogDismiss(props);
  return _VKYTJKYNcjs.createElement.call(void 0, "button", htmlProps);
});
if (process.env.NODE_ENV !== "production") {
  DialogDismiss.displayName = "DialogDismiss";
}




exports.useDialogDismiss = useDialogDismiss; exports.DialogDismiss = DialogDismiss;
