// @ariakit/react-core/cjs/__chunks/KXLF6YSN.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});"use client";



var _FQGSRSYDcjs = require('./FQGSRSYD.cjs');



var _KZA6P235cjs = require('./KZA6P235.cjs');


var _VKYTJKYNcjs = require('./VKYTJKYN.cjs');

// src/combobox/combobox-context.tsx
var _react = require('react');
var ctx = _VKYTJKYNcjs.createStoreContext.call(void 0, 
  [_KZA6P235cjs.PopoverContextProvider, _FQGSRSYDcjs.CompositeContextProvider],
  [_KZA6P235cjs.PopoverScopedContextProvider, _FQGSRSYDcjs.CompositeScopedContextProvider]
);
var useComboboxContext = ctx.useContext;
var useComboboxScopedContext = ctx.useScopedContext;
var useComboboxProviderContext = ctx.useProviderContext;
var ComboboxContextProvider = ctx.ContextProvider;
var ComboboxScopedContextProvider = ctx.ScopedContextProvider;
var ComboboxItemValueContext = _react.createContext.call(void 0, 
  void 0
);








exports.useComboboxContext = useComboboxContext; exports.useComboboxScopedContext = useComboboxScopedContext; exports.useComboboxProviderContext = useComboboxProviderContext; exports.ComboboxContextProvider = ComboboxContextProvider; exports.ComboboxScopedContextProvider = ComboboxScopedContextProvider; exports.ComboboxItemValueContext = ComboboxItemValueContext;
