// @ariakit/react-core/cjs/__chunks/L5APETYY.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});"use client";


var _YREZ6S36cjs = require('./YREZ6S36.cjs');


var _D5XLZTLOcjs = require('./D5XLZTLO.cjs');


var _DJLIJTDUcjs = require('./DJLIJTDU.cjs');


var _JF225FQ5cjs = require('./JF225FQ5.cjs');



var _VV6WA3I6cjs = require('./VV6WA3I6.cjs');



var _DEZWSLIMcjs = require('./DEZWSLIM.cjs');


var _PZL34OVOcjs = require('./PZL34OVO.cjs');


var _5GTNIPQ6cjs = require('./5GTNIPQ6.cjs');


var _PRN5A3P4cjs = require('./PRN5A3P4.cjs');


var _G3CEEQUPcjs = require('./G3CEEQUP.cjs');


var _J3OG6T3Bcjs = require('./J3OG6T3B.cjs');


var _AE6TH3T4cjs = require('./AE6TH3T4.cjs');



var _RXGXI75Ncjs = require('./RXGXI75N.cjs');





var _QKKT6RXXcjs = require('./QKKT6RXX.cjs');


var _7ARL6QHNcjs = require('./7ARL6QHN.cjs');




var _VKYTJKYNcjs = require('./VKYTJKYN.cjs');


var _QQNI2PXZcjs = require('./QQNI2PXZ.cjs');


var _WLJVTO7Acjs = require('./WLJVTO7A.cjs');








var _JTVNRMACcjs = require('./JTVNRMAC.cjs');




var _WYXAWBLEcjs = require('./WYXAWBLE.cjs');

// src/dialog/dialog.tsx
var _react = require('react');






var _dom = require('@ariakit/core/utils/dom');



var _events = require('@ariakit/core/utils/events');




var _focus = require('@ariakit/core/utils/focus');
var _misc = require('@ariakit/core/utils/misc');
var _platform = require('@ariakit/core/utils/platform');
var _jsxruntime = require('react/jsx-runtime');
var isSafariBrowser = _platform.isSafari.call(void 0, );
function isAlreadyFocusingAnotherElement(dialog) {
  const activeElement = _dom.getActiveElement.call(void 0, );
  if (!activeElement)
    return false;
  if (dialog && _dom.contains.call(void 0, dialog, activeElement))
    return false;
  if (_focus.isFocusable.call(void 0, activeElement))
    return true;
  return false;
}
function getElementFromProp(prop, focusable = false) {
  if (!prop)
    return null;
  const element = "current" in prop ? prop.current : prop;
  if (!element)
    return null;
  if (focusable)
    return _focus.isFocusable.call(void 0, element) ? element : null;
  return element;
}
var useDialog = _VKYTJKYNcjs.createHook.call(void 0, 
  (_a) => {
    var _b = _a, {
      store: storeProp,
      open: openProp,
      onClose,
      focusable = true,
      modal = true,
      portal = !!modal,
      backdrop = !!modal,
      backdropProps,
      hideOnEscape = true,
      hideOnInteractOutside = true,
      getPersistentElements,
      preventBodyScroll = !!modal,
      autoFocusOnShow = true,
      autoFocusOnHide = true,
      initialFocus,
      finalFocus,
      unmountOnHide
    } = _b, props = _WYXAWBLEcjs.__objRest.call(void 0, _b, [
      "store",
      "open",
      "onClose",
      "focusable",
      "modal",
      "portal",
      "backdrop",
      "backdropProps",
      "hideOnEscape",
      "hideOnInteractOutside",
      "getPersistentElements",
      "preventBodyScroll",
      "autoFocusOnShow",
      "autoFocusOnHide",
      "initialFocus",
      "finalFocus",
      "unmountOnHide"
    ]);
    const context = _QKKT6RXXcjs.useDialogProviderContext.call(void 0, );
    const ref = _react.useRef.call(void 0, null);
    const store = _QQNI2PXZcjs.useDialogStore.call(void 0, {
      store: storeProp || context,
      open: openProp,
      setOpen(open2) {
        if (open2)
          return;
        const dialog = ref.current;
        if (!dialog)
          return;
        const event = new Event("close", { bubbles: false, cancelable: true });
        if (onClose) {
          dialog.addEventListener("close", onClose, { once: true });
        }
        dialog.dispatchEvent(event);
        if (!event.defaultPrevented)
          return;
        store.setOpen(true);
      }
    });
    const { portalRef, domReady } = _JTVNRMACcjs.usePortalRef.call(void 0, portal, props.portalRef);
    const preserveTabOrderProp = props.preserveTabOrder;
    const preserveTabOrder = store.useState(
      (state) => preserveTabOrderProp && !modal && state.mounted
    );
    const id = _JTVNRMACcjs.useId.call(void 0, props.id);
    const open = store.useState("open");
    const mounted = store.useState("mounted");
    const contentElement = store.useState("contentElement");
    const hidden = _RXGXI75Ncjs.isHidden.call(void 0, mounted, props.hidden, props.alwaysVisible);
    _PRN5A3P4cjs.usePreventBodyScroll.call(void 0, contentElement, id, preventBodyScroll && !hidden);
    _DJLIJTDUcjs.useHideOnInteractOutside.call(void 0, store, hideOnInteractOutside, domReady);
    const { wrapElement, nestedDialogs } = _D5XLZTLOcjs.useNestedDialogs.call(void 0, store);
    props = _JTVNRMACcjs.useWrapElement.call(void 0, props, wrapElement, [wrapElement]);
    if (process.env.NODE_ENV !== "production") {
      _react.useEffect.call(void 0, () => {
        if (!backdropProps)
          return;
        console.warn(
          "The `backdropProps` prop is deprecated. Use the `backdrop` prop instead.",
          "See https://ariakit.org/reference/dialog#backdrop"
        );
      }, [backdropProps]);
    }
    _JTVNRMACcjs.useSafeLayoutEffect.call(void 0, () => {
      if (!open)
        return;
      const dialog = ref.current;
      const activeElement = _dom.getActiveElement.call(void 0, dialog, true);
      if (!activeElement)
        return;
      if (activeElement.tagName === "BODY")
        return;
      if (dialog && _dom.contains.call(void 0, dialog, activeElement))
        return;
      store.setDisclosureElement(activeElement);
    }, [store, open]);
    if (isSafariBrowser) {
      _react.useEffect.call(void 0, () => {
        if (!mounted)
          return;
        const { disclosureElement } = store.getState();
        if (!disclosureElement)
          return;
        if (!_dom.isButton.call(void 0, disclosureElement))
          return;
        const onMouseDown = () => {
          let receivedFocus = false;
          const onFocus = () => {
            receivedFocus = true;
          };
          const options = { capture: true, once: true };
          disclosureElement.addEventListener("focusin", onFocus, options);
          _events.queueBeforeEvent.call(void 0, disclosureElement, "mouseup", () => {
            disclosureElement.removeEventListener("focusin", onFocus, true);
            if (receivedFocus)
              return;
            _focus.focusIfNeeded.call(void 0, disclosureElement);
          });
        };
        disclosureElement.addEventListener("mousedown", onMouseDown);
        return () => {
          disclosureElement.removeEventListener("mousedown", onMouseDown);
        };
      }, [store, mounted]);
    }
    const shouldDisableAccessibilityTree = modal || // Usually, we only want to disable the accessibility tree outside if the
    // dialog is a modal. But the Portal component can't preserve the tab
    // order on Safari/VoiceOver. By allowing only the dialog/portal to be
    // accessible, we provide a similar tab order flow. We don't need to
    // disable pointer events because it's just for screen readers.
    portal && preserveTabOrder && _platform.isSafari.call(void 0, );
    _react.useEffect.call(void 0, () => {
      if (!mounted)
        return;
      if (!domReady)
        return;
      const dialog = ref.current;
      if (!dialog)
        return;
      if (!shouldDisableAccessibilityTree)
        return;
      const existingDismiss = dialog.querySelector("[data-dialog-dismiss]");
      if (existingDismiss)
        return;
      return _JF225FQ5cjs.prependHiddenDismiss.call(void 0, dialog, store.hide);
    }, [store, mounted, domReady, shouldDisableAccessibilityTree]);
    _JTVNRMACcjs.useSafeLayoutEffect.call(void 0, () => {
      if (open)
        return;
      if (!mounted)
        return;
      if (!domReady)
        return;
      const dialog = ref.current;
      if (!dialog)
        return;
      return _DEZWSLIMcjs.disableTree.call(void 0, dialog);
    }, [open, mounted, domReady]);
    const canTakeTreeSnapshot = open && domReady;
    _JTVNRMACcjs.useSafeLayoutEffect.call(void 0, () => {
      if (!id)
        return;
      if (!canTakeTreeSnapshot)
        return;
      const dialog = ref.current;
      return _5GTNIPQ6cjs.createWalkTreeSnapshot.call(void 0, id, [dialog]);
    }, [id, canTakeTreeSnapshot]);
    const getPersistentElementsProp = _JTVNRMACcjs.useEvent.call(void 0, getPersistentElements);
    _JTVNRMACcjs.useSafeLayoutEffect.call(void 0, () => {
      if (!id)
        return;
      if (!canTakeTreeSnapshot)
        return;
      const { disclosureElement } = store.getState();
      const dialog = ref.current;
      const persistentElements = getPersistentElementsProp() || [];
      const allElements = [
        dialog,
        ...persistentElements,
        ...nestedDialogs.map((dialog2) => dialog2.getState().contentElement)
      ];
      if (!shouldDisableAccessibilityTree) {
        return _VV6WA3I6cjs.markTreeOutside.call(void 0, id, [disclosureElement, ...allElements]);
      }
      if (modal) {
        return _misc.chain.call(void 0, 
          _VV6WA3I6cjs.markTreeOutside.call(void 0, id, allElements),
          _DEZWSLIMcjs.disableTreeOutside.call(void 0, id, allElements)
        );
      }
      return _misc.chain.call(void 0, 
        _VV6WA3I6cjs.markTreeOutside.call(void 0, id, [disclosureElement, ...allElements]),
        _PZL34OVOcjs.disableAccessibilityTreeOutside.call(void 0, id, allElements)
      );
    }, [
      id,
      store,
      canTakeTreeSnapshot,
      getPersistentElementsProp,
      nestedDialogs,
      shouldDisableAccessibilityTree,
      modal
    ]);
    const mayAutoFocusOnShow = !!autoFocusOnShow;
    const autoFocusOnShowProp = _JTVNRMACcjs.useBooleanEvent.call(void 0, autoFocusOnShow);
    const [autoFocusEnabled, setAutoFocusEnabled] = _react.useState.call(void 0, false);
    _react.useEffect.call(void 0, () => {
      if (!open)
        return;
      if (!mayAutoFocusOnShow)
        return;
      if (!domReady)
        return;
      if (!(contentElement == null ? void 0 : contentElement.isConnected))
        return;
      const element = getElementFromProp(initialFocus, true) || // If no initial focus is specified, we try to focus the first element
      // with the autofocus attribute. If it's an Ariakit component, the
      // Focusable component will consume the autoFocus prop and add the
      // data-autofocus attribute to the element instead.
      contentElement.querySelector(
        "[data-autofocus=true],[autofocus]"
      ) || // We have to fallback to the first focusable element otherwise portaled
      // dialogs with preserveTabOrder set to true will not receive focus
      // properly because the elements aren't tabbable until the dialog
      // receives focus.
      _focus.getFirstTabbableIn.call(void 0, contentElement, true, portal && preserveTabOrder) || // Finally, we fallback to the dialog element itself.
      contentElement;
      const isElementFocusable = _focus.isFocusable.call(void 0, element);
      if (!autoFocusOnShowProp(isElementFocusable ? element : null))
        return;
      setAutoFocusEnabled(true);
      queueMicrotask(() => {
        element.focus();
        if (!isSafariBrowser)
          return;
        element.scrollIntoView({ block: "nearest", inline: "nearest" });
      });
    }, [
      open,
      mayAutoFocusOnShow,
      domReady,
      contentElement,
      initialFocus,
      portal,
      preserveTabOrder,
      autoFocusOnShowProp
    ]);
    const mayAutoFocusOnHide = !!autoFocusOnHide;
    const autoFocusOnHideProp = _JTVNRMACcjs.useBooleanEvent.call(void 0, autoFocusOnHide);
    const [hasOpened, setHasOpened] = _react.useState.call(void 0, false);
    _react.useEffect.call(void 0, () => {
      if (!open)
        return;
      setHasOpened(true);
      return () => setHasOpened(false);
    }, [open]);
    const focusOnHide = _react.useCallback.call(void 0, 
      (dialog, retry = true) => {
        const { disclosureElement } = store.getState();
        if (isAlreadyFocusingAnotherElement(dialog))
          return;
        let element = getElementFromProp(finalFocus) || disclosureElement;
        if (element == null ? void 0 : element.id) {
          const doc = _dom.getDocument.call(void 0, element);
          const selector = `[aria-activedescendant="${element.id}"]`;
          const composite = doc.querySelector(selector);
          if (composite) {
            element = composite;
          }
        }
        if (element && !_focus.isFocusable.call(void 0, element)) {
          const maybeParentDialog = _dom.closest.call(void 0, element, "[data-dialog]");
          if (maybeParentDialog && maybeParentDialog.id) {
            const doc = _dom.getDocument.call(void 0, maybeParentDialog);
            const selector = `[aria-controls~="${maybeParentDialog.id}"]`;
            const control = doc.querySelector(selector);
            if (control) {
              element = control;
            }
          }
        }
        const isElementFocusable = element && _focus.isFocusable.call(void 0, element);
        if (!isElementFocusable && retry) {
          requestAnimationFrame(() => focusOnHide(dialog, false));
          return;
        }
        if (!autoFocusOnHideProp(isElementFocusable ? element : null))
          return;
        if (!isElementFocusable)
          return;
        element == null ? void 0 : element.focus();
      },
      [store, finalFocus, autoFocusOnHideProp]
    );
    _JTVNRMACcjs.useSafeLayoutEffect.call(void 0, () => {
      if (open)
        return;
      if (!hasOpened)
        return;
      if (!mayAutoFocusOnHide)
        return;
      const dialog = ref.current;
      focusOnHide(dialog);
    }, [open, hasOpened, domReady, mayAutoFocusOnHide, focusOnHide]);
    _react.useEffect.call(void 0, () => {
      if (!hasOpened)
        return;
      if (!mayAutoFocusOnHide)
        return;
      const dialog = ref.current;
      return () => focusOnHide(dialog);
    }, [hasOpened, mayAutoFocusOnHide, focusOnHide]);
    const hideOnEscapeProp = _JTVNRMACcjs.useBooleanEvent.call(void 0, hideOnEscape);
    _react.useEffect.call(void 0, () => {
      if (!domReady)
        return;
      if (!mounted)
        return;
      const onKeyDown = (event) => {
        if (event.key !== "Escape")
          return;
        if (event.defaultPrevented)
          return;
        const dialog = ref.current;
        if (!dialog)
          return;
        if (_VV6WA3I6cjs.isElementMarked.call(void 0, dialog))
          return;
        const target = event.target;
        if (!target)
          return;
        const { disclosureElement } = store.getState();
        const isValidTarget = () => {
          if (target.tagName === "BODY")
            return true;
          if (_dom.contains.call(void 0, dialog, target))
            return true;
          if (!disclosureElement)
            return true;
          if (_dom.contains.call(void 0, disclosureElement, target))
            return true;
          return false;
        };
        if (!isValidTarget())
          return;
        if (!hideOnEscapeProp(event))
          return;
        store.hide();
      };
      return _events.addGlobalEventListener.call(void 0, "keydown", onKeyDown, true);
    }, [store, domReady, mounted, hideOnEscapeProp]);
    props = _JTVNRMACcjs.useWrapElement.call(void 0, 
      props,
      (element) => /* @__PURE__ */ _jsxruntime.jsx.call(void 0, _J3OG6T3Bcjs.HeadingLevel, { level: modal ? 1 : void 0, children: element }),
      [modal]
    );
    const hiddenProp = props.hidden;
    const alwaysVisible = props.alwaysVisible;
    props = _JTVNRMACcjs.useWrapElement.call(void 0, 
      props,
      (element) => {
        if (!backdrop)
          return element;
        return /* @__PURE__ */ _jsxruntime.jsxs.call(void 0, _jsxruntime.Fragment, { children: [
          /* @__PURE__ */ _jsxruntime.jsx.call(void 0, 
            _YREZ6S36cjs.DialogBackdrop,
            {
              store,
              backdrop,
              backdropProps,
              hidden: hiddenProp,
              alwaysVisible
            }
          ),
          element
        ] });
      },
      [store, backdrop, backdropProps, hiddenProp, alwaysVisible]
    );
    const [headingId, setHeadingId] = _react.useState.call(void 0, );
    const [descriptionId, setDescriptionId] = _react.useState.call(void 0, );
    props = _JTVNRMACcjs.useWrapElement.call(void 0, 
      props,
      (element) => /* @__PURE__ */ _jsxruntime.jsx.call(void 0, _QKKT6RXXcjs.DialogScopedContextProvider, { value: store, children: /* @__PURE__ */ _jsxruntime.jsx.call(void 0, _QKKT6RXXcjs.DialogHeadingContext.Provider, { value: setHeadingId, children: /* @__PURE__ */ _jsxruntime.jsx.call(void 0, _QKKT6RXXcjs.DialogDescriptionContext.Provider, { value: setDescriptionId, children: element }) }) }),
      [store]
    );
    props = _WYXAWBLEcjs.__spreadProps.call(void 0, _WYXAWBLEcjs.__spreadValues.call(void 0, {
      id,
      "data-dialog": "",
      role: "dialog",
      tabIndex: focusable ? -1 : void 0,
      "aria-labelledby": headingId,
      "aria-describedby": descriptionId
    }, props), {
      ref: _JTVNRMACcjs.useMergeRefs.call(void 0, ref, props.ref)
    });
    props = _G3CEEQUPcjs.useFocusableContainer.call(void 0, _WYXAWBLEcjs.__spreadProps.call(void 0, _WYXAWBLEcjs.__spreadValues.call(void 0, {}, props), {
      autoFocusOnShow: autoFocusEnabled
    }));
    props = _RXGXI75Ncjs.useDisclosureContent.call(void 0, _WYXAWBLEcjs.__spreadValues.call(void 0, { store }, props));
    props = _7ARL6QHNcjs.useFocusable.call(void 0, _WYXAWBLEcjs.__spreadProps.call(void 0, _WYXAWBLEcjs.__spreadValues.call(void 0, {}, props), { focusable }));
    props = _AE6TH3T4cjs.usePortal.call(void 0, _WYXAWBLEcjs.__spreadProps.call(void 0, _WYXAWBLEcjs.__spreadValues.call(void 0, { portal }, props), { portalRef, preserveTabOrder }));
    return props;
  }
);
function createDialogComponent(Component, useProviderContext = _QKKT6RXXcjs.useDialogProviderContext) {
  return _VKYTJKYNcjs.createComponent.call(void 0, (props) => {
    const context = useProviderContext();
    const store = props.store || context;
    const mounted = _WLJVTO7Acjs.useStoreState.call(void 0, 
      store,
      (state) => !props.unmountOnHide || (state == null ? void 0 : state.mounted) || !!props.open
    );
    if (!mounted)
      return null;
    return /* @__PURE__ */ _jsxruntime.jsx.call(void 0, Component, _WYXAWBLEcjs.__spreadValues.call(void 0, {}, props));
  });
}
var Dialog = createDialogComponent(
  _VKYTJKYNcjs.createComponent.call(void 0, (props) => {
    const htmlProps = useDialog(props);
    return _VKYTJKYNcjs.createElement.call(void 0, "div", htmlProps);
  }),
  _QKKT6RXXcjs.useDialogProviderContext
);
if (process.env.NODE_ENV !== "production") {
  Dialog.displayName = "Dialog";
}





exports.useDialog = useDialog; exports.createDialogComponent = createDialogComponent; exports.Dialog = Dialog;
