// @ariakit/react-core/cjs/__chunks/M3F5A3H5.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});"use client";



var _FQGSRSYDcjs = require('./FQGSRSYD.cjs');



var _KZA6P235cjs = require('./KZA6P235.cjs');


var _VKYTJKYNcjs = require('./VKYTJKYN.cjs');

// src/select/select-context.tsx
var _react = require('react');
var ctx = _VKYTJKYNcjs.createStoreContext.call(void 0, 
  [_KZA6P235cjs.PopoverContextProvider, _FQGSRSYDcjs.CompositeContextProvider],
  [_KZA6P235cjs.PopoverScopedContextProvider, _FQGSRSYDcjs.CompositeScopedContextProvider]
);
var useSelectContext = ctx.useContext;
var useSelectScopedContext = ctx.useScopedContext;
var useSelectProviderContext = ctx.useProviderContext;
var SelectContextProvider = ctx.ContextProvider;
var SelectScopedContextProvider = ctx.ScopedContextProvider;
var SelectItemCheckedContext = _react.createContext.call(void 0, false);








exports.useSelectContext = useSelectContext; exports.useSelectScopedContext = useSelectScopedContext; exports.useSelectProviderContext = useSelectProviderContext; exports.SelectContextProvider = SelectContextProvider; exports.SelectScopedContextProvider = SelectScopedContextProvider; exports.SelectItemCheckedContext = SelectItemCheckedContext;
