// @ariakit/react-core/cjs/__chunks/PG7QUWN4.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});"use client";


var _QCAPQDE7cjs = require('./QCAPQDE7.cjs');




var _VKYTJKYNcjs = require('./VKYTJKYN.cjs');



var _WYXAWBLEcjs = require('./WYXAWBLE.cjs');

// src/focus-trap/focus-trap.ts
var useFocusTrap = _VKYTJKYNcjs.createHook.call(void 0, (props) => {
  props = _WYXAWBLEcjs.__spreadProps.call(void 0, _WYXAWBLEcjs.__spreadValues.call(void 0, {
    "data-focus-trap": "",
    tabIndex: 0,
    "aria-hidden": true
  }, props), {
    style: _WYXAWBLEcjs.__spreadValues.call(void 0, {
      // Prevents unintended scroll jumps.
      position: "fixed",
      top: 0,
      left: 0
    }, props.style)
  });
  props = _QCAPQDE7cjs.useVisuallyHidden.call(void 0, props);
  return props;
});
var FocusTrap = _VKYTJKYNcjs.createComponent.call(void 0, (props) => {
  const htmlProps = useFocusTrap(props);
  return _VKYTJKYNcjs.createElement.call(void 0, "span", htmlProps);
});
if (process.env.NODE_ENV !== "production") {
  FocusTrap.displayName = "FocusTrap";
}




exports.useFocusTrap = useFocusTrap; exports.FocusTrap = FocusTrap;
