// @ariakit/react-core/cjs/__chunks/QCAPQDE7.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});"use client";




var _VKYTJKYNcjs = require('./VKYTJKYN.cjs');



var _WYXAWBLEcjs = require('./WYXAWBLE.cjs');

// src/visually-hidden/visually-hidden.ts
var useVisuallyHidden = _VKYTJKYNcjs.createHook.call(void 0, (props) => {
  props = _WYXAWBLEcjs.__spreadProps.call(void 0, _WYXAWBLEcjs.__spreadValues.call(void 0, {}, props), {
    style: _WYXAWBLEcjs.__spreadValues.call(void 0, {
      border: 0,
      clip: "rect(0 0 0 0)",
      height: "1px",
      margin: "-1px",
      overflow: "hidden",
      padding: 0,
      position: "absolute",
      whiteSpace: "nowrap",
      width: "1px"
    }, props.style)
  });
  return props;
});
var VisuallyHidden = _VKYTJKYNcjs.createComponent.call(void 0, 
  (props) => {
    const htmlProps = useVisuallyHidden(props);
    return _VKYTJKYNcjs.createElement.call(void 0, "span", htmlProps);
  }
);
if (process.env.NODE_ENV !== "production") {
  VisuallyHidden.displayName = "VisuallyHidden";
}




exports.useVisuallyHidden = useVisuallyHidden; exports.VisuallyHidden = VisuallyHidden;
