// @ariakit/react-core/cjs/__chunks/QKKT6RXX.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});"use client";



var _SMKW5GKScjs = require('./SMKW5GKS.cjs');


var _VKYTJKYNcjs = require('./VKYTJKYN.cjs');

// src/dialog/dialog-context.tsx
var _react = require('react');
var ctx = _VKYTJKYNcjs.createStoreContext.call(void 0, 
  [_SMKW5GKScjs.DisclosureContextProvider],
  [_SMKW5GKScjs.DisclosureScopedContextProvider]
);
var useDialogContext = ctx.useContext;
var useDialogScopedContext = ctx.useScopedContext;
var useDialogProviderContext = ctx.useProviderContext;
var DialogContextProvider = ctx.ContextProvider;
var DialogScopedContextProvider = ctx.ScopedContextProvider;
var DialogHeadingContext = _react.createContext.call(void 0, void 0);
var DialogDescriptionContext = _react.createContext.call(void 0, void 0);









exports.useDialogContext = useDialogContext; exports.useDialogScopedContext = useDialogScopedContext; exports.useDialogProviderContext = useDialogProviderContext; exports.DialogContextProvider = DialogContextProvider; exports.DialogScopedContextProvider = DialogScopedContextProvider; exports.DialogHeadingContext = DialogHeadingContext; exports.DialogDescriptionContext = DialogDescriptionContext;
