// @ariakit/react-core/cjs/__chunks/YRDDHZT7.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});"use client";



var _FQGSRSYDcjs = require('./FQGSRSYD.cjs');



var _ORT5QOHHcjs = require('./ORT5QOHH.cjs');


var _VKYTJKYNcjs = require('./VKYTJKYN.cjs');

// src/menu/menu-context.ts
var _react = require('react');
var menu = _VKYTJKYNcjs.createStoreContext.call(void 0, 
  [_FQGSRSYDcjs.CompositeContextProvider, _ORT5QOHHcjs.HovercardContextProvider],
  [_FQGSRSYDcjs.CompositeScopedContextProvider, _ORT5QOHHcjs.HovercardScopedContextProvider]
);
var useMenuContext = menu.useContext;
var useMenuScopedContext = menu.useScopedContext;
var useMenuProviderContext = menu.useProviderContext;
var MenuContextProvider = menu.ContextProvider;
var MenuScopedContextProvider = menu.ScopedContextProvider;
var menubar = _VKYTJKYNcjs.createStoreContext.call(void 0, 
  [_FQGSRSYDcjs.CompositeContextProvider],
  [_FQGSRSYDcjs.CompositeScopedContextProvider]
);
var useMenuBarContext = menubar.useContext;
var useMenuBarScopedContext = menubar.useScopedContext;
var useMenuBarProviderContext = menubar.useProviderContext;
var MenuBarContextProvider = menubar.ContextProvider;
var MenuBarScopedContextProvider = menubar.ScopedContextProvider;
var MenuItemCheckedContext = _react.createContext.call(void 0, 
  void 0
);













exports.useMenuContext = useMenuContext; exports.useMenuScopedContext = useMenuScopedContext; exports.useMenuProviderContext = useMenuProviderContext; exports.MenuContextProvider = MenuContextProvider; exports.MenuScopedContextProvider = MenuScopedContextProvider; exports.useMenuBarContext = useMenuBarContext; exports.useMenuBarScopedContext = useMenuBarScopedContext; exports.useMenuBarProviderContext = useMenuBarProviderContext; exports.MenuBarContextProvider = MenuBarContextProvider; exports.MenuBarScopedContextProvider = MenuBarScopedContextProvider; exports.MenuItemCheckedContext = MenuItemCheckedContext;
