// @ariakit/react-core/cjs/__chunks/YREZ6S36.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});"use client";


var _VV6WA3I6cjs = require('./VV6WA3I6.cjs');


var _2TEJFTKYcjs = require('./2TEJFTKY.cjs');


var _RXGXI75Ncjs = require('./RXGXI75N.cjs');


var _4H55S5YCcjs = require('./4H55S5YC.cjs');



var _JTVNRMACcjs = require('./JTVNRMAC.cjs');



var _WYXAWBLEcjs = require('./WYXAWBLE.cjs');

// src/dialog/dialog-backdrop.tsx
var _react = require('react');
var _jsxruntime = require('react/jsx-runtime');
function DialogBackdrop({
  store,
  backdrop,
  backdropProps,
  alwaysVisible,
  hidden
}) {
  const ref = _react.useRef.call(void 0, null);
  const disclosure = _4H55S5YCcjs.useDisclosureStore.call(void 0, { disclosure: store });
  const contentElement = store.useState("contentElement");
  _JTVNRMACcjs.useSafeLayoutEffect.call(void 0, () => {
    const backdrop2 = ref.current;
    const dialog = contentElement;
    if (!backdrop2)
      return;
    if (!dialog)
      return;
    backdrop2.style.zIndex = getComputedStyle(dialog).zIndex;
  }, [contentElement]);
  _JTVNRMACcjs.useSafeLayoutEffect.call(void 0, () => {
    const id = contentElement == null ? void 0 : contentElement.id;
    if (!id)
      return;
    const backdrop2 = ref.current;
    if (!backdrop2)
      return;
    return _VV6WA3I6cjs.markAncestor.call(void 0, backdrop2, id);
  }, [contentElement]);
  if (hidden != null) {
    backdropProps = _WYXAWBLEcjs.__spreadProps.call(void 0, _WYXAWBLEcjs.__spreadValues.call(void 0, {}, backdropProps), { hidden });
  }
  const props = _RXGXI75Ncjs.useDisclosureContent.call(void 0, _WYXAWBLEcjs.__spreadProps.call(void 0, _WYXAWBLEcjs.__spreadValues.call(void 0, {
    store: disclosure,
    role: "presentation",
    "data-backdrop": (contentElement == null ? void 0 : contentElement.id) || "",
    alwaysVisible
  }, backdropProps), {
    ref: _JTVNRMACcjs.useMergeRefs.call(void 0, backdropProps == null ? void 0 : backdropProps.ref, ref),
    style: _WYXAWBLEcjs.__spreadValues.call(void 0, {
      position: "fixed",
      top: 0,
      right: 0,
      bottom: 0,
      left: 0
    }, backdropProps == null ? void 0 : backdropProps.style)
  }));
  if (!backdrop)
    return null;
  if (_react.isValidElement.call(void 0, backdrop)) {
    return /* @__PURE__ */ _jsxruntime.jsx.call(void 0, _2TEJFTKYcjs.Role, _WYXAWBLEcjs.__spreadProps.call(void 0, _WYXAWBLEcjs.__spreadValues.call(void 0, {}, props), { render: backdrop }));
  }
  const Component = typeof backdrop !== "boolean" ? backdrop : "div";
  return /* @__PURE__ */ _jsxruntime.jsx.call(void 0, _2TEJFTKYcjs.Role, _WYXAWBLEcjs.__spreadProps.call(void 0, _WYXAWBLEcjs.__spreadValues.call(void 0, {}, props), { render: /* @__PURE__ */ _jsxruntime.jsx.call(void 0, Component, {}) }));
}



exports.DialogBackdrop = DialogBackdrop;
