// @ariakit/core/cjs/__chunks/LUOEAZGU.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});"use client";


var _ROX3MRGOcjs = require('./ROX3MRGO.cjs');


var _OASSVCYRcjs = require('./OASSVCYR.cjs');


var _LMT335NJcjs = require('./LMT335NJ.cjs');



var _WYXAWBLEcjs = require('./WYXAWBLE.cjs');

// src/hovercard/hovercard-store.ts
function createHovercardStore(props = {}) {
  var _a;
  const syncState = (_a = props.store) == null ? void 0 : _a.getState();
  const popover = _ROX3MRGOcjs.createPopoverStore.call(void 0, _WYXAWBLEcjs.__spreadProps.call(void 0, _WYXAWBLEcjs.__spreadValues.call(void 0, {}, props), {
    placement: _LMT335NJcjs.defaultValue.call(void 0, 
      props.placement,
      syncState == null ? void 0 : syncState.placement,
      "bottom"
    )
  }));
  const timeout = _LMT335NJcjs.defaultValue.call(void 0, props.timeout, syncState == null ? void 0 : syncState.timeout, 500);
  const initialState = _WYXAWBLEcjs.__spreadProps.call(void 0, _WYXAWBLEcjs.__spreadValues.call(void 0, {}, popover.getState()), {
    timeout,
    showTimeout: _LMT335NJcjs.defaultValue.call(void 0, props.showTimeout, syncState == null ? void 0 : syncState.showTimeout),
    hideTimeout: _LMT335NJcjs.defaultValue.call(void 0, props.hideTimeout, syncState == null ? void 0 : syncState.hideTimeout),
    autoFocusOnShow: _LMT335NJcjs.defaultValue.call(void 0, syncState == null ? void 0 : syncState.autoFocusOnShow, false)
  });
  const hovercard = _OASSVCYRcjs.createStore.call(void 0, initialState, popover, props.store);
  return _WYXAWBLEcjs.__spreadProps.call(void 0, _WYXAWBLEcjs.__spreadValues.call(void 0, _WYXAWBLEcjs.__spreadValues.call(void 0, {}, popover), hovercard), {
    setAutoFocusOnShow: (value) => hovercard.setState("autoFocusOnShow", value)
  });
}



exports.createHovercardStore = createHovercardStore;
