// @ariakit/core/cjs/__chunks/ROX3MRGO.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});"use client";


var _MLPOXRDTcjs = require('./MLPOXRDT.cjs');





var _OASSVCYRcjs = require('./OASSVCYR.cjs');


var _LMT335NJcjs = require('./LMT335NJ.cjs');




var _WYXAWBLEcjs = require('./WYXAWBLE.cjs');

// src/popover/popover-store.ts
function createPopoverStore(_a = {}) {
  var _b = _a, {
    popover: otherPopover
  } = _b, props = _WYXAWBLEcjs.__objRest.call(void 0, _b, [
    "popover"
  ]);
  const store = _OASSVCYRcjs.mergeStore.call(void 0, 
    props.store,
    _OASSVCYRcjs.omit.call(void 0, otherPopover, [
      "arrowElement",
      "anchorElement",
      "contentElement",
      "popoverElement",
      "disclosureElement"
    ])
  );
  _OASSVCYRcjs.throwOnConflictingProps.call(void 0, props, store);
  const syncState = store == null ? void 0 : store.getState();
  const dialog = _MLPOXRDTcjs.createDialogStore.call(void 0, _WYXAWBLEcjs.__spreadProps.call(void 0, _WYXAWBLEcjs.__spreadValues.call(void 0, {}, props), { store }));
  const placement = _LMT335NJcjs.defaultValue.call(void 0, 
    props.placement,
    syncState == null ? void 0 : syncState.placement,
    "bottom"
  );
  const initialState = _WYXAWBLEcjs.__spreadProps.call(void 0, _WYXAWBLEcjs.__spreadValues.call(void 0, {}, dialog.getState()), {
    placement,
    currentPlacement: placement,
    anchorElement: _LMT335NJcjs.defaultValue.call(void 0, syncState == null ? void 0 : syncState.anchorElement, null),
    popoverElement: _LMT335NJcjs.defaultValue.call(void 0, syncState == null ? void 0 : syncState.popoverElement, null),
    arrowElement: _LMT335NJcjs.defaultValue.call(void 0, syncState == null ? void 0 : syncState.arrowElement, null),
    rendered: Symbol("rendered")
  });
  const popover = _OASSVCYRcjs.createStore.call(void 0, initialState, dialog, store);
  return _WYXAWBLEcjs.__spreadProps.call(void 0, _WYXAWBLEcjs.__spreadValues.call(void 0, _WYXAWBLEcjs.__spreadValues.call(void 0, {}, dialog), popover), {
    setAnchorElement: (element) => popover.setState("anchorElement", element),
    setPopoverElement: (element) => popover.setState("popoverElement", element),
    setArrowElement: (element) => popover.setState("arrowElement", element),
    render: () => popover.setState("rendered", Symbol("rendered"))
  });
}



exports.createPopoverStore = createPopoverStore;
