// @ariakit/core/cjs/__chunks/YXSGST3H.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});"use client";








var _OASSVCYRcjs = require('./OASSVCYR.cjs');


var _LMT335NJcjs = require('./LMT335NJ.cjs');



var _WYXAWBLEcjs = require('./WYXAWBLE.cjs');

// src/disclosure/disclosure-store.ts
function createDisclosureStore(props = {}) {
  const store = _OASSVCYRcjs.mergeStore.call(void 0, 
    props.store,
    _OASSVCYRcjs.omit.call(void 0, props.disclosure, ["contentElement", "disclosureElement"])
  );
  _OASSVCYRcjs.throwOnConflictingProps.call(void 0, props, store);
  const syncState = store == null ? void 0 : store.getState();
  const open = _LMT335NJcjs.defaultValue.call(void 0, 
    props.open,
    syncState == null ? void 0 : syncState.open,
    props.defaultOpen,
    false
  );
  const animated = _LMT335NJcjs.defaultValue.call(void 0, props.animated, syncState == null ? void 0 : syncState.animated, false);
  const initialState = {
    open,
    animated,
    animating: !!animated && open,
    mounted: open,
    contentElement: _LMT335NJcjs.defaultValue.call(void 0, syncState == null ? void 0 : syncState.contentElement, null),
    disclosureElement: _LMT335NJcjs.defaultValue.call(void 0, syncState == null ? void 0 : syncState.disclosureElement, null)
  };
  const disclosure = _OASSVCYRcjs.createStore.call(void 0, initialState, store);
  _OASSVCYRcjs.setup.call(void 0, 
    disclosure,
    () => _OASSVCYRcjs.sync.call(void 0, disclosure, ["animated", "animating"], (state) => {
      if (state.animated)
        return;
      disclosure.setState("animating", false);
    })
  );
  _OASSVCYRcjs.setup.call(void 0, 
    disclosure,
    () => _OASSVCYRcjs.subscribe.call(void 0, disclosure, ["open"], () => {
      if (!disclosure.getState().animated)
        return;
      disclosure.setState("animating", true);
    })
  );
  _OASSVCYRcjs.setup.call(void 0, 
    disclosure,
    () => _OASSVCYRcjs.sync.call(void 0, disclosure, ["open", "animating"], (state) => {
      disclosure.setState("mounted", state.open || state.animating);
    })
  );
  return _WYXAWBLEcjs.__spreadProps.call(void 0, _WYXAWBLEcjs.__spreadValues.call(void 0, {}, disclosure), {
    setOpen: (value) => disclosure.setState("open", value),
    show: () => disclosure.setState("open", true),
    hide: () => disclosure.setState("open", false),
    toggle: () => disclosure.setState("open", (open2) => !open2),
    stopAnimation: () => disclosure.setState("animating", false),
    setContentElement: (value) => disclosure.setState("contentElement", value),
    setDisclosureElement: (value) => disclosure.setState("disclosureElement", value)
  });
}



exports.createDisclosureStore = createDisclosureStore;
