// @wordpress/components/build/box-control/unit-control.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = BoxUnitControl;
var _react = require("react");
var _react2 = require("@use-gesture/react");
var _tooltip = _interopRequireDefault(require("../tooltip"));
var _boxControlStyles = require("./styles/box-control-styles");
/**
 * External dependencies
 */

/**
 * Internal dependencies
 */

const noop = () => {};
function BoxUnitControl({
  isFirst,
  isLast,
  isOnly,
  onHoverOn = noop,
  onHoverOff = noop,
  label,
  value,
  ...props
}) {
  const bindHoverGesture = (0, _react2.useHover)(({
    event,
    ...state
  }) => {
    if (state.hovering) {
      onHoverOn(event, state);
    } else {
      onHoverOff(event, state);
    }
  });
  return (0, _react.createElement)(_boxControlStyles.UnitControlWrapper, {
    ...bindHoverGesture()
  }, (0, _react.createElement)(Tooltip, {
    text: label
  }, (0, _react.createElement)(_boxControlStyles.UnitControl, {
    "aria-label": label,
    className: "component-box-control__unit-control",
    isFirst: isFirst,
    isLast: isLast,
    isOnly: isOnly,
    isPressEnterToChange: true,
    isResetValueOnUnitChange: false,
    value: value,
    ...props
  })));
}
function Tooltip({
  children,
  text
}) {
  if (!text) return children;

  /**
   * Wrapping the children in a `<div />` as Tooltip as it attempts
   * to render the <UnitControl />. Using a plain `<div />` appears to
   * resolve this issue.
   *
   * Originally discovered and referenced here:
   * https://github.com/WordPress/gutenberg/pull/24966#issuecomment-685875026
   */
  return (0, _react.createElement)(_tooltip.default, {
    text: text,
    placement: "top"
  }, (0, _react.createElement)("div", null, children));
}
