// @wordpress/components/build/private-apis.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.unlock = exports.privateApis = exports.lock = void 0;
var _privateApis = require("@wordpress/private-apis");
var _v = require("./composite/v2");
var _customSelectControl = _interopRequireDefault(require("./custom-select-control"));
var _utils = require("./popover/utils");
var _progressBar = _interopRequireDefault(require("./progress-bar"));
var _slotFill = require("./slot-fill");
var _dropdownMenuV = require("./dropdown-menu-v2");
var _dropdownMenuV2Ariakit = require("./dropdown-menu-v2-ariakit");
var _contextSystemProvider = require("./context/context-system-provider");
var _theme = _interopRequireDefault(require("./theme"));
var _tabs = _interopRequireDefault(require("./tabs"));
/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */

const {
  lock,
  unlock
} = (0, _privateApis.__dangerousOptInToUnstableAPIsOnlyForCoreModules)('I know using unstable features means my theme or plugin will inevitably break in the next version of WordPress.', '@wordpress/components');
exports.unlock = unlock;
exports.lock = lock;
const privateApis = {};
exports.privateApis = privateApis;
lock(privateApis, {
  CompositeV2: _v.Composite,
  CompositeGroupV2: _v.CompositeGroup,
  CompositeItemV2: _v.CompositeItem,
  CompositeRowV2: _v.CompositeRow,
  useCompositeStoreV2: _v.useCompositeStore,
  CustomSelectControl: _customSelectControl.default,
  __experimentalPopoverLegacyPositionToPlacement: _utils.positionToPlacement,
  createPrivateSlotFill: _slotFill.createPrivateSlotFill,
  ComponentsContext: _contextSystemProvider.ComponentsContext,
  DropdownMenuV2: _dropdownMenuV.DropdownMenu,
  DropdownMenuCheckboxItemV2: _dropdownMenuV.DropdownMenuCheckboxItem,
  DropdownMenuGroupV2: _dropdownMenuV.DropdownMenuGroup,
  DropdownMenuItemV2: _dropdownMenuV.DropdownMenuItem,
  DropdownMenuLabelV2: _dropdownMenuV.DropdownMenuLabel,
  DropdownMenuRadioGroupV2: _dropdownMenuV.DropdownMenuRadioGroup,
  DropdownMenuRadioItemV2: _dropdownMenuV.DropdownMenuRadioItem,
  DropdownMenuSeparatorV2: _dropdownMenuV.DropdownMenuSeparator,
  DropdownSubMenuV2: _dropdownMenuV.DropdownSubMenu,
  DropdownSubMenuTriggerV2: _dropdownMenuV.DropdownSubMenuTrigger,
  ProgressBar: _progressBar.default,
  Tabs: _tabs.default,
  Theme: _theme.default,
  DropdownMenuV2Ariakit: _dropdownMenuV2Ariakit.DropdownMenu,
  DropdownMenuGroupV2Ariakit: _dropdownMenuV2Ariakit.DropdownMenuGroup,
  DropdownMenuGroupLabelV2Ariakit: _dropdownMenuV2Ariakit.DropdownMenuGroupLabel,
  DropdownMenuItemV2Ariakit: _dropdownMenuV2Ariakit.DropdownMenuItem,
  DropdownMenuCheckboxItemV2Ariakit: _dropdownMenuV2Ariakit.DropdownMenuCheckboxItem,
  DropdownMenuRadioItemV2Ariakit: _dropdownMenuV2Ariakit.DropdownMenuRadioItem,
  DropdownMenuSeparatorV2Ariakit: _dropdownMenuV2Ariakit.DropdownMenuSeparator
});
