// @wordpress/components/build/tabs/tab.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Tab = void 0;
var _react = require("react");
var _element = require("@wordpress/element");
var _warning = _interopRequireDefault(require("@wordpress/warning"));
var _context = require("./context");
var _styles = require("./styles");
/**
 * WordPress dependencies
 */

const Tab = (0, _element.forwardRef)(function Tab({
  children,
  id,
  className,
  disabled,
  render,
  style
}, ref) {
  const context = (0, _element.useContext)(_context.TabsContext);
  if (!context) {
    typeof SCRIPT_DEBUG !== "undefined" && SCRIPT_DEBUG === true ? (0, _warning.default)('`Tabs.TabList` must be wrapped in a `Tabs` component.') : void 0;
    return null;
  }
  const {
    store,
    instanceId
  } = context;
  const instancedTabId = `${instanceId}-${id}`;
  return (0, _react.createElement)(_styles.Tab, {
    ref: ref,
    store: store,
    id: instancedTabId,
    className: className,
    style: style,
    disabled: disabled,
    render: render
  }, children);
});
exports.Tab = Tab;
