// @ariakit/react-core/cjs/select/select-item.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});"use client";


var _QWH4QBHAcjs = require('../__chunks/QWH4QBHA.cjs');



var _M3F5A3H5cjs = require('../__chunks/M3F5A3H5.cjs');


var _H2TTW65Qcjs = require('../__chunks/H2TTW65Q.cjs');
require('../__chunks/F7HB6RYG.cjs');
require('../__chunks/HY2NPBHW.cjs');
require('../__chunks/RV3PHHWM.cjs');
require('../__chunks/FQGSRSYD.cjs');
require('../__chunks/JVORNAMD.cjs');
require('../__chunks/KZA6P235.cjs');
require('../__chunks/QKKT6RXX.cjs');
require('../__chunks/SMKW5GKS.cjs');
require('../__chunks/7ARL6QHN.cjs');
require('../__chunks/75KXQZJX.cjs');




var _VKYTJKYNcjs = require('../__chunks/VKYTJKYN.cjs');
require('../__chunks/WLJVTO7A.cjs');





var _JTVNRMACcjs = require('../__chunks/JTVNRMAC.cjs');
require('../__chunks/LNE2FWJO.cjs');




var _WYXAWBLEcjs = require('../__chunks/WYXAWBLE.cjs');

// src/select/select-item.tsx
var _react = require('react');
var _dom = require('@ariakit/core/utils/dom');
var _events = require('@ariakit/core/utils/events');
var _misc = require('@ariakit/core/utils/misc');
var _jsxruntime = require('react/jsx-runtime');
function isSelected(storeValue, itemValue) {
  if (storeValue == null)
    return false;
  if (itemValue == null)
    return false;
  if (Array.isArray(storeValue)) {
    return storeValue.includes(itemValue);
  }
  return storeValue === itemValue;
}
var useSelectItem = _VKYTJKYNcjs.createHook.call(void 0, 
  (_a) => {
    var _b = _a, {
      store,
      value,
      getItem: getItemProp,
      hideOnClick,
      setValueOnClick = value != null,
      preventScrollOnKeyDown = true,
      focusOnHover = true
    } = _b, props = _WYXAWBLEcjs.__objRest.call(void 0, _b, [
      "store",
      "value",
      "getItem",
      "hideOnClick",
      "setValueOnClick",
      "preventScrollOnKeyDown",
      "focusOnHover"
    ]);
    var _a2;
    const context = _M3F5A3H5cjs.useSelectScopedContext.call(void 0, );
    store = store || context;
    _misc.invariant.call(void 0, 
      store,
      process.env.NODE_ENV !== "production" && "SelectItem must be wrapped in a SelectList or SelectPopover component."
    );
    const id = _JTVNRMACcjs.useId.call(void 0, props.id);
    const disabled = _misc.disabledFromProps.call(void 0, props);
    const getItem = _react.useCallback.call(void 0, 
      (item) => {
        const nextItem = _WYXAWBLEcjs.__spreadProps.call(void 0, _WYXAWBLEcjs.__spreadValues.call(void 0, {}, item), {
          value: disabled ? void 0 : value,
          children: value
        });
        if (getItemProp) {
          return getItemProp(nextItem);
        }
        return nextItem;
      },
      [disabled, value, getItemProp]
    );
    const multiSelectable = store.useState(
      (state) => Array.isArray(state.value)
    );
    hideOnClick = hideOnClick != null ? hideOnClick : value != null && !multiSelectable;
    const onClickProp = props.onClick;
    const setValueOnClickProp = _JTVNRMACcjs.useBooleanEvent.call(void 0, setValueOnClick);
    const hideOnClickProp = _JTVNRMACcjs.useBooleanEvent.call(void 0, hideOnClick);
    const onClick = _JTVNRMACcjs.useEvent.call(void 0, (event) => {
      onClickProp == null ? void 0 : onClickProp(event);
      if (event.defaultPrevented)
        return;
      if (_events.isDownloading.call(void 0, event))
        return;
      if (_events.isOpeningInNewTab.call(void 0, event))
        return;
      if (setValueOnClickProp(event) && value != null) {
        store == null ? void 0 : store.setValue((prevValue) => {
          if (!Array.isArray(prevValue))
            return value;
          if (prevValue.includes(value)) {
            return prevValue.filter((v) => v !== value);
          }
          return [...prevValue, value];
        });
      }
      if (hideOnClickProp(event)) {
        store == null ? void 0 : store.hide();
      }
    });
    const selected = store.useState((state) => isSelected(state.value, value));
    props = _JTVNRMACcjs.useWrapElement.call(void 0, 
      props,
      (element) => /* @__PURE__ */ _jsxruntime.jsx.call(void 0, _M3F5A3H5cjs.SelectItemCheckedContext.Provider, { value: selected, children: element }),
      [selected]
    );
    const contentElement = store.useState("contentElement");
    const autoFocus = store.useState((state) => {
      if (state.activeId !== id && (store == null ? void 0 : store.item(state.activeId)))
        return false;
      if (state.value == null)
        return false;
      if (value == null)
        return false;
      if (Array.isArray(state.value)) {
        return state.value[state.value.length - 1] === value;
      }
      return state.value === value;
    });
    props = _WYXAWBLEcjs.__spreadProps.call(void 0, _WYXAWBLEcjs.__spreadValues.call(void 0, {
      id,
      role: _dom.getPopupItemRole.call(void 0, contentElement),
      "aria-selected": selected,
      children: value
    }, props), {
      autoFocus: (_a2 = props.autoFocus) != null ? _a2 : autoFocus,
      onClick
    });
    props = _H2TTW65Qcjs.useCompositeItem.call(void 0, _WYXAWBLEcjs.__spreadValues.call(void 0, {
      store,
      getItem,
      preventScrollOnKeyDown
    }, props));
    const focusOnHoverProp = _JTVNRMACcjs.useBooleanEvent.call(void 0, focusOnHover);
    props = _QWH4QBHAcjs.useCompositeHover.call(void 0, _WYXAWBLEcjs.__spreadProps.call(void 0, _WYXAWBLEcjs.__spreadValues.call(void 0, {
      store
    }, props), {
      // We have to disable focusOnHover when the popup is closed, otherwise
      // the active item will change to null (the container) when the popup is
      // closed by clicking on an item.
      focusOnHover: (event) => {
        if (!focusOnHoverProp(event))
          return false;
        const state = store == null ? void 0 : store.getState();
        return !!(state == null ? void 0 : state.open);
      }
    }));
    return props;
  }
);
var SelectItem = _VKYTJKYNcjs.createMemoComponent.call(void 0, (props) => {
  const htmlProps = useSelectItem(props);
  return _VKYTJKYNcjs.createElement.call(void 0, "div", htmlProps);
});
if (process.env.NODE_ENV !== "production") {
  SelectItem.displayName = "SelectItem";
}



exports.SelectItem = SelectItem; exports.useSelectItem = useSelectItem;
